src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -193,51 +220,49 @@ internal ItemManager(RecentItemsManager recentItemsManager)
                         filterText.Length <= 1)
                     {
-                        initialListOfItemsToBeIncluded.Add(new ExtendedFilterResult(item, new FilterResult(roslynItem, filterText, matchedFilterText: false)));
+                        builder.Add(new ExtendedFilterResult(item, new FilterResult(roslynItem, filterText, matchedFilterText: false, patternMatch)));
                     }
                 }
             }
 
-            // DismissIfLastCharacterDeleted should be applied only when started with Insertion, and then Deleted all characters typed.
-            // This confirms with the original VS 2010 behavior.
-            if (initialRoslynTriggerKind == CompletionTriggerKind.Insertion &&
-                data.Trigger.Reason == CompletionTriggerReason.Backspace &&
-                completionRules.DismissIfLastCharacterDeleted &&
-                session.ApplicableToSpan.GetText(data.Snapshot).Length == 0)
-            {
-                // Dismiss the session
-                return null;
-            }
-
-            if (initialListOfItemsToBeIncluded.Count == 0)
+            if (builder.Count == 0)
             {
                 return HandleAllItemsFilteredOut(reason, data.SelectedFilters, completionRules);
             }
 
-            var options = document?.Project.Solution.Options;
-            var highlightMatchingPortions = options?.GetOption(CompletionOptions.HighlightMatchingPortionsOfCompletionListItems, document.Project.Language) ?? true;
+            var experimentationService = document?.Project.Solution.Workspace.Services.GetService<IExperimentationService>();
+
+            var initialListOfItemsToBeIncluded = experimentationService?.IsExperimentEnabled(WellKnownExperimentNames.SortCompletionListByMatch) == true
+                // Need a stable sorting algorithm to preserve the original order for items with same pattern match score.
+                ? builder.OrderBy(new NullablePatternMatchComparer()).ToImmutableArray()
+                : builder.ToImmutable();
+
+            builder.Free();
+
             var showCompletionItemFilters = options?.GetOption(CompletionOptions.ShowCompletionItemFilters, document.Project.Language) ?? true;
 
             var updatedFilters = showCompletionItemFilters
                 ? GetUpdatedFilters(initialListOfItemsToBeIncluded, data.SelectedFilters)
                 : ImmutableArray<CompletionFilterWithState>.Empty;
 
-            var highlightedList = GetHighlightedList(initialListOfItemsToBeIncluded, filterText, completionHelper, highlightMatchingPortions).ToImmutableArray();
+            var highlightedList = highlightMatchingPortions
+                ? GetHighlightedList(initialListOfItemsToBeIncluded, filterText, completionHelper)
+                : initialListOfItemsToBeIncluded.SelectAsArray(r => new CompletionItemWithHighlight(r.VSCompletionItem, ImmutableArray<Span>.Empty));
```

Comment:
if there's no actua perf hit to getting the matches, then i think we should just always do this.  then, the code is so much cleaner because we don't need all these different pathsthat have to check this value.

Reply:
With this change, it shows insignificant difference with highlight on and off, about 40 out of ~1400ms for the tested scenario. I'd consider this once I have done some more comprehensive perf test.